patches/api/0203-Spawn-Reason-API.patch

```diff
@@ -22,33 +22,130 @@ index a89fff5c164881be0286ec2240e94dd5883ecc40..e55f6e2baf35dbd91c433ab9e62713ea
+      * @return Arrow entity spawned as a result of this method
+      */
+     @NotNull
+-    public <T extends AbstractArrow> T spawnArrow(@NotNull Location location, @NotNull Vector direction, float speed, float spread, @NotNull Class<T> clazz);
++    // Paper start
++    default  <T extends AbstractArrow> T spawnArrow(@NotNull Location location, @NotNull Vector direction, float speed, float spread, @NotNull Class<T> clazz) {
++        return spawnArrow(location, direction, speed, spread, clazz, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT); // should be custom but bukkit do that
++    }
++
++    /**
++     * Creates an arrow entity of the given class at the given {@link Location}
++     *
++     * @param <T> type of arrow to spawn
++     * @param location Location to spawn the arrow
++     * @param direction Direction to shoot the arrow in
++     * @param speed Speed of the arrow. A recommend speed is 0.6
++     * @param spread Spread of the arrow. A recommend spread is 12
++     * @param clazz the Entity class for the arrow
++     * {@link org.bukkit.entity.SpectralArrow SpectralArrow},{@link org.bukkit.entity.Arrow Arrow},{@link org.bukkit.entity.TippedArrow TippedArrow}
++     * @param reason the spawn reason
++     * @return Arrow entity spawned as a result of this method
++     */
++    @NotNull
++    <T extends AbstractArrow> T spawnArrow(@NotNull Location location, @NotNull Vector direction, float speed, float spread, @NotNull Class<T> clazz, @NotNull org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason);
++    // Paper end
+ 
+     /**
+      * Creates a tree at the given {@link Location}
+@@ -738,7 +758,34 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
+      * @return The lightning entity.
+      */
+     @NotNull
+-    public LightningStrike strikeLightning(@NotNull Location loc);
++    // Paper start
++    default LightningStrike strikeLightning(@NotNull Location loc) {
++        return strikeLightning(loc, true);
++    }
++
++    /**
++     * Strikes lightning at the given {@link Location}
++     *
++     * @param loc The location to strike lightning
++     * @param doDamage Weather or not this lightning should do damage
++     * @return The lightning entity.
++     */
++    @NotNull
++    default LightningStrike strikeLightning(@NotNull Location loc, boolean doDamage) {
++        return strikeLightning(loc, doDamage, org.bukkit.event.weather.LightningStrikeEvent.Cause.CUSTOM);
++    }
++
++    /**
++     * Strikes lightning at the given {@link Location}
++     *
++     * @param loc The location to strike lightning
++     * @param doDamage Weather or not this lightning should do damage
++     * @param cause The cause of this lightning
++     * @return The lightning entity.
++     */
++    @NotNull
++    LightningStrike strikeLightning(@NotNull Location loc, boolean doDamage, @NotNull org.bukkit.event.weather.LightningStrikeEvent.Cause cause);
```

Comment:
I am unsure about the general "doDamage" parameter, That is what `strikeLightningEffect` is there for.
I see how a single unified method might be nicer, but those methods have existed for ages.

Reply:
Yeah i will revert, it was just obvious for me